[mc] measurement & warmup timers

HugoStrand opened the pull request:
Here is an updated pullrequest.

This is a fix for issue #334 adding two separate timers for warmup and accumulation, and separate timers for the measurements. This is particularly useful for monitoring the two-particle measurements that dominate the accumulation.

I have also added timestamps to the percentage printout and an "estimated time of arrival" (ETA) giving output like:

```
Welcome to the CTHYB solver

╔╦╗╦═╗╦╔═╗ ╔═╗  ┌─┐┌┬┐┬ ┬┬ ┬┌┐ 
 ║ ╠╦╝║║═╬╗╚═╗  │   │ ├─┤└┬┘├┴┐
 ╩ ╩╚═╩╚═╝╚╚═╝  └─┘ ┴ ┴ ┴ ┴ └─┘

The local Hamiltonian of the problem:
-2*c_dag('down',0)*c('down',0) + -2*c_dag('up',0)*c('up',0)
Using autopartition algorithm to partition the local Hilbert space
Found 4 subspaces.

Warming up ...
12:13:43  21% ETA 00:00:01 n_cycles, NC = 1000, 216

Accumulating ...
12:13:43   0% ETA 00:00:13 n_cycles, NC = 500, 4
12:13:45  31% ETA 00:00:05 n_cycles, NC = 500, 156
12:13:48  71% ETA 00:00:02 n_cycles, NC = 500, 355


[Rank 0] Timings for all measures:
Measure                                           | seconds   
Average sign                                      | 0.000159321
G2_iw fermionic measurement                       | 0.0724332 
G2_iw_ph particle-hole measurement                | 0.305697  
G2_iw_pp particle-particle measurement            | 0.303132  
G2_iwll_pp Legendre particle-particle measurement | 5.34491   
G2_tau imaginary-time measurement                 | 0.0387911 
G_tau measure                                     | 0.000896209
Total measure time                                | 6.06601   
[Rank 0] Acceptance rate for all moves:
Move set Insert two operators: 0.108525
  Move  Insert Delta_down: 0.106569
  Move  Insert Delta_up: 0.110488
Move set Remove two operators: 0.107223
  Move  Remove Delta_down: 0.106303
  Move  Remove Delta_up: 0.108126
Move  Shift one operator: 0.681289
[Rank 0] Warmup lasted: 0.407273 seconds [00:00:01]
[Rank 0] Simulation lasted: 6.27942 seconds [00:00:07]
[Rank 0] Number of measures: 500
Total number of measures: 500
Average sign: 1
```

HugoStrand commented:
Dear @parcollet and @Wentzell could you merge or refute this? Or may I merge it?

I keep on rebasing for a month now, to keep this on top of my own unstable...

Best, Hugo

HugoStrand commented:
@parcollet @Wentzell if I do not hear back on this from you I will merge tomorrow myself. Best, Hugo

mferrero commented:
Hi Hugo. I had a look at your pull request. I have a couple of comments:

- Do we really want Timer to be an argument of the run method? Are there situations when the user will want to provide his own timer? It looks to me like we could use a single timer in the run method and copy the elapsed time in two separate variables according to the value of do_measure. This would also avoid to break the API.

- Cosmetics: I think it is nicer if we have a more uniform style for variables and have timer rather than Timer or timer_acc rather than TimerAcc. We never really use CamelCase in triqs I think.

- Maybe it would be nice to have methods to get the different run times. Maybe get_warmup_time(), get_accumulation_time() and get_total_time()? And we keep get_duration() for backward compat? What do you think @parcollet, @Wentzell, could that be useful?

Cheers!

parcollet commented:
I agree.
We were also looking at this this morning with Nils.
@Wentzell : you can summarise our remarks, they are in line with Michel's.

PS : for the variable, I agree of course. That we not the case probably in this code 
for internal variables since it is a bit old, but it should be.

HugoStrand commented:
Hello @mferrero,

Thank you for the feedback! I have fixed the points you raise, please see the updated pull request.

On the side of CamelCase there are some examples in `mc_generic.hpp`, e.g.
```c++
  random_generator RandomGenerator;
  move_set<MCSignType> AllMoves;
  measure_set<MCSignType> AllMeasures;
  std::vector<measure_aux> AllMeasuresAux;
```
 I'll try to stay clear of that in the future.

Cheers, Hugo

HugoStrand commented on triqs/mc_tools/mc_generic.hpp:
```diff
@@ -226,18 +230,27 @@ namespace mc_tools {
+     report << utility::timestamp() << " "
+	    << std::setfill(' ') << std::setw(3) << done_percent << "%"
+            << " ETA " << estimate_time_left(n_cycles, NC, timer)
+            << " n_cycles, NC = " << n_cycles << ", " << NC 
```
Goot point, fixed in 63f2574

HugoStrand commented on triqs/utility/timer_chrono.hpp:
```diff
@@ -0,0 +1,59 @@
+
```
My reasoning for separating the timer from the other time code is that the `timer_chrono.hpp` uses a high resolution clock which we might want to change to a different type in the future.

Wentzell replied:
Our point was that triqs does not need both a highres and a lowres timer but rather just one timer class.

HugoStrand replied:
Ok! Good.

I do not know enough about the dependencies of the previous timer to dare to remove it, while at the same time I need the hi-res timer to get ok numbers on the very fast measurements.

By adding a hi-res one rather than modifying the existing one,  I am not breaking any old code. Is that ok?

HugoStrand commented on triqs/utility/timer_chrono.hpp:
```diff
@@ -0,0 +1,59 @@
+
+    class timer {
+      public:
+      typedef std::chrono::high_resolution_clock clock_t;
```
Fixed in 63f2574

HugoStrand commented on triqs/utility/timer_chrono.hpp:
```diff
@@ -0,0 +1,59 @@
+      public:
+      typedef std::chrono::high_resolution_clock clock_t;
+      typedef clock_t::time_point time_point_t;
+      typedef clock_t::duration duration_t;
```
Thanks for spotting this. Fixed in 63f2574

HugoStrand commented on triqs/utility/timer_chrono.hpp:
```diff
@@ -0,0 +1,59 @@
+      bool is_running() const { return running; }
+      operator double() const {
+        std::chrono::duration<double> total_time_seconds(total_time);
+	if(is_running()) total_time_seconds += clock_t::now() - start_time;
```
Fixed in 63f2574

HugoStrand commented on triqs/utility/timestamp.hpp:
```diff
@@ -0,0 +1,54 @@
+
```
I would like to separate the timer class from the general code dealing with time stamps etc. In order to easily be able to change the timer type if needed

Wentzell replied:
See comment regarding single timer class

HugoStrand commented on triqs/utility/timestamp.hpp:
```diff
@@ -0,0 +1,54 @@
+
+    std::string inline hours_minutes_seconds_from_seconds(double seconds) {
+      using namespace std::chrono;
+      time_point<system_clock> tp(milliseconds(long(seconds * 1000)) - hours(1));
```
Long story.. Turns out that the `time_point` is initiated with the given time plus one hour. I do not know why but this produces the expected output.

Wentzell replied:
Somehow it seems to have swallowed my other comment. 
I think this function is unclear in what it does, using system_clock and localtime
only to convert to number of seconds into hours/minutes/seconds. 
Further it will be problematic if used for runtimes longer than a day.

I suggest to change it to
```c++
std::string hours_minutes_seconds_from_seconds(double sec)
{
    using namespace std::chrono;
    using days = duration<int, std::ratio<86400>>;
    auto s = seconds{int(sec)}; 
    auto d = duration_cast<days>(s);
    auto h = duration_cast<hours>(s -= d);
    auto m = duration_cast<minutes>(s -= h);
    s -= m; 
    std::ostringstream sstream;
    sstream.fill('0');
    sstream << std::setw(2) << d.count() << "d:"
       << std::setw(2) << h.count() << "h:"
       << std::setw(2) << m.count() << "m:"
       << std::setw(2) << s.count() << 's';
    return sstream.str(); 
}
```

HugoStrand replied:
I am sorry but I disagree. 

The function name and call signature describe exactly what it does.

It solves the problem using the standard library routines for handling time, instead of re-implementing existing functionality. I am strongly against re-implementing time conversion by hand, and against re-implementing the time conversion to string, as you suggest above.

For multiple-day cases the number of hours should simply be more than 24, as one is used to in working with HPC queuing systems. I want the format as HHH:MM:SS since this is the direct input one has to give for wall-time to most HPC schedulers.

In the end I think this is a matter of opinion.

Wentzell replied:
> Long story.. Turns out that the time_point is initiated with the given time plus one hour. I do not know why but this produces the expected output.

I tried your solution on wandbox.org and it gives a different result than on my local machine. Subtracting one hour seems to be a hardcoded solution depending on the localtime of your machine.

Wentzell replied:
> The function name and call signature describe exactly what it does.

We want both the signature **and** the code to be clear. From what I found that functionality does not exist, so we should be explicit. Your solution tries to use the functionality formatting clocks to, while we should consider std::chrono::duration.
There does not seem to be a formatting tool for that.

> For multiple-day cases the number of hours should simply be more than 24, as one is used to in working with HPC queuing systems. I want the format as HHH:MM:SS since this is the direct input one has to give for wall-time to most HPC schedulers.

I agree that format might be a better choice. However the put_time will just reset to 0 Hours if you go past 24.

HugoStrand commented on triqs/utility/timestamp.hpp:
```diff
@@ -0,0 +1,54 @@
+      return s.str();
+    }
+    
+    std::string inline estimate_time_left(int Ntot, int iter, timer & Timer) {
```
Fair comment.

I was trying to make the additions to `mc_generic.hpp` as simple as possible, and factoring out this logic makes it easier to read that code. If we where to incorporate this code in the `mc_generic` run method there would be rather verbose call to `hours_minutes_seconds_from_seconds(...)` in that method. I argue that it is easier to read `estimate_time_left(...)`, which tells you exactly what is going on. It does not even require a comment on what is happening since the function name is self-explanatory.

HugoStrand commented:
@Wentzell thank you for the review comments. Please have a look at my changes and consider approving the pull request. Best, H

Wentzell commented on triqs/utility/timer_chrono.hpp:
```diff
@@ -0,0 +1,57 @@
+      bool running;
+
+      public:
+      timer() : total_time(0), running(false) {}
```
Remove it. It will be defaulted.

Wentzell commented on triqs/utility/timer_chrono.hpp:
```diff
@@ -0,0 +1,57 @@
+      private:
+      clock_t::time_point start_time;
+      clock_t::duration total_time;
+      bool running;
```
```c++
clock_t::duration total_time = 0;
bool running = false;
```

HugoStrand commented:
Dear @Wentzell 

I have replaced the `timer.hpp` with what used to be `timer_chrono.hpp` and removed the explicit constructor. Can we now merge?

Best, H
